drivers/lora/shell.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) 2020 Andreas Sandberg
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/lora.h>
+#include <inttypes.h>
+#include <shell/shell.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+
+LOG_MODULE_REGISTER(lora_shell);
+
+#define DEFAULT_RADIO_NODE DT_ALIAS(lora0)
+BUILD_ASSERT(DT_HAS_NODE(DEFAULT_RADIO_NODE),
+	     "No default LoRa radio specified in DT");
+#define DEFAULT_RADIO DT_LABEL(DEFAULT_RADIO_NODE)
+
+static struct lora_modem_config modem_config = {
+	.frequency = 0,
+	.bandwidth = BW_125_KHZ,
+	.datarate = SF_10,
+	.coding_rate = CR_4_5,
+	.preamble_len = 8,
+	.tx_power = 4,
+};
+
+static const int bw_table[] = {
+	[BW_125_KHZ] = 125,
+	[BW_250_KHZ] = 250,
+	[BW_500_KHZ] = 500,
+};
+
+static struct device *get_modem(const struct shell *shell,
+				struct lora_modem_config *cfg)
+{
+	struct device *dev;
+	int ret;
+
+	dev = device_get_binding(DEFAULT_RADIO);
+	if (!dev) {
+		shell_error(shell, "%s Device not found", DEFAULT_RADIO);
+		return NULL;
+	}
+
+	if (cfg) {
+		if (cfg->frequency == 0) {
+			shell_error(shell, "No frequency specified.");
+			return NULL;
+		}
+
+		ret = lora_config(dev, cfg);
+		if (ret < 0) {
+			shell_error(shell, "LoRa config failed");
+			return NULL;
+		}
+	}
+
+	return dev;
+}
+static int cmd_lora_conf(const struct shell *shell, size_t argc, char **argv)
+{
+	if (argc < 2) {
+		shell_print(shell, DEFAULT_RADIO ":");
+		shell_print(shell, "  Frequency: %" PRIu32 " Hz",
+			    modem_config.frequency);
+		shell_print(shell, "  TX power: %" PRIi8 " dBm",
+			    modem_config.tx_power);
+		shell_print(shell, "  Bandwidth: %i kHz",
+			    bw_table[modem_config.bandwidth]);
+		shell_print(shell, "  Spreading factor: SF%i",
+			    (int)modem_config.datarate);
+		shell_print(shell, "  Coding rate: 4/%i",
+			    (int)modem_config.coding_rate + 4);
+		shell_print(shell, "  Preamble length: %" PRIu16,
+			    modem_config.preamble_len);
+
+		return 0;
+	}
+
+	for (int i = 1; i < argc; i += 2) {
+		char *param = argv[i];
+		char *eptr;
+		int value;
```

Comment:
What about introducing some helper function like `lora_conf_param_set(const struct shell *shell, const char *param, const char *value)` and then calling `lora_conf_param_set(shell, argv[i], argv[i + 1]);` in the loop? Currently there are too many lines of code and too many local variables in this function.